accepted/single-file/extract.md

```diff
@@ -0,0 +1,81 @@
+
+* Extract to a location within the temp-directory, so that if some files are not removed (ex: because an app crashed before cleanup), they are removed when the temp-directory is purged.
+* Be short, to reduce the risk of running over path length limit.
+* Be randomized, in order to support concurrent launches of the same app.
+
+The proposed extraction path is `%TEMP\.net\<random-code>` / `$TMPDIR/.net/<random-code>` . 
+
+#### Pros
+
+* Simpler implementation: fewer complexities with respect to app-updates and concurrency.
+* User need not worry about explicit cleanup.
+* Some customers don't prefer a click-and-run model, rather than an app-install model. This option better suits them.
+
+#### Cons
+
+* Higher startup cost, because of extraction on every run -- especially for self-contained apps.
+
+Mono and Costura extract native dependencies to temporary files on every run.
+
+## Persistent Extraction
+
+The requirements for persistent extraction are:
+
+* Reuse: Extract on first-run, reuse on subsequent runs.
+* Fault-tolerance: Recover from failure after partial extraction on first run.
+* Concurrency: Handle concurrent launches on first start.
+* Upgrade: Each version of the app extracts to a unique location, supporting side-by-side use of multiple versions.
+* Uninstall: Users can identify and delete extracted files when the app is no longer needed.
+* Access control: Processes running with elevated access can extract to admin-only-writable locations.
+
+At first startup, the host takes a disk lock (in order to handle concurrent launches), extracts files to an *install-location*, verifies them, and writes and *end-marker* before releasing the lock. If the host fails to extract all the resources at startup, it will attempt to clean up all contents of the install-location before termination. 
+
+Subsequent runs reuse the extracted files after confirming their successful extraction. However, if the contents of install-location are corrupted post-extraction, they will need to be explicitly cleaned.
+
+By default, the host extracts dependencies to `BASE_PATH/APP_NAME/ID/PERM/`
+
+- `BASE_PATH` is  `%HOMEPATH%\.dotnet\AppDependencies ` /  `$HOME/.dotnet/AppDependencies`
+- `APP_NAME` is simple the name of the app (the host executable containing embedded data).
+- `ID` is a hash-code based on the host binary to distinguish app version and architecture specifications.
+- `PERM` is `user` for normal runs, and `admin` for elevated runs with appropriate write-permissions.
+
+A different extraction location may be specified via a `runtimeconfig` option.
+
+The cleanup of extracted files in the install-location will be manual in this version. We can consider adding `dotnet CLI` commands for cleanup in future. However, future versions are expected to spill fewer artifacts to disk, making the cleanup commands a lower priority feature.
```

Comment:
For self-contained apps, users are not expected to have a dotnet CLI, so there might be some other approach needed to clean it up. Would it be possible for the app to clean up itself, to implement some *“Exit and uninstall”* functionality into the application?

Reply:
You mean not cleanup on every exit, but on a specific uninstall run, right? Yes, the app-developers can implement this, because the install-location of extracted files is predictable.

Reply:
Yeah, using a persistent location (so that the extracting only happens once) but at the same time offering an option _within_ the application to remove the extracted files at that location.